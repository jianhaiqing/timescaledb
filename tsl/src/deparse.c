/*
 * This file and its contents are licensed under the Timescale License.
 * Please see the included NOTICE for copyright information and
 * LICENSE-TIMESCALE for a copy of the license.
 */
#include <postgres.h>
#include <utils/rel.h>
#include <lib/stringinfo.h>
#include <utils/builtins.h>
#include <utils/lsyscache.h>
#include <utils/relcache.h>
#include <catalog/indexing.h>
#include <utils/ruleutils.h>
#include <utils/syscache.h>
#include <commands/tablespace.h>
#include <catalog/pg_class.h>
#include <utils/rel.h>
#include <access/relscan.h>
#include <utils/fmgroids.h>
#include <catalog/pg_constraint.h>
#include <catalog/pg_index.h>
#include <nodes/pg_list.h>
#include <fmgr.h>

#include "export.h"
#include "compat.h"
#include "constraint.h"
#include "trigger.h"
#include "utils.h"
#include "deparse.h"
#include <extension.h>
#include <utils/lsyscache.h>

/*
 * Deparse a table into a set of SQL commands that can be used to recreate it.
 * Together with column definiton it deparses constraints, indexes, triggers
 * and rules as well.  There are some table types that are not supported:
 * temporary, partitioned, foreign, inherited and a table that uses
 * options. Row security is also not supported.
 */
typedef const char *(*GetCmdFunc)(Oid oid);

static const char *
get_index_cmd(Oid oid)
{
	return pg_get_indexdef_string(oid);
}

static const char *
get_constraint_cmd(Oid oid)
{
	return pg_get_constraintdef_command(oid);
}

static FunctionCallInfo
build_fcinfo_data(Oid oid)
{
	FunctionCallInfo fcinfo = palloc(sizeof(SizeForFunctionCallInfo(1)));

	InitFunctionCallInfoData(*fcinfo, NULL, 1, InvalidOid, NULL, NULL);
	FC_ARG(fcinfo, 0) = ObjectIdGetDatum(oid);
	FC_NULL(fcinfo, 0) = false;

	return fcinfo;
}

static const char *
get_trigger_cmd(Oid oid)
{
	return TextDatumGetCString(pg_get_triggerdef(build_fcinfo_data(oid)));
}

static const char *
get_rule_cmd(Oid oid)
{
	return TextDatumGetCString(pg_get_ruledef(build_fcinfo_data(oid)));
}

static List *
get_cmds(List *oids, GetCmdFunc get_cmd)
{
	List *cmds = NIL;
	ListCell *cell;

	foreach (cell, oids)
	{
		StringInfo cmd = makeStringInfo();

		appendStringInfo(cmd, "%s;", get_cmd(lfirst_oid(cell)));
		cmds = lappend(cmds, cmd->data);
	}
	return cmds;
}

static List *
get_constraint_cmds(List *constraint_oids)
{
	return get_cmds(constraint_oids, get_constraint_cmd);
}

static List *
get_index_cmds(List *index_oids)
{
	return get_cmds(index_oids, get_index_cmd);
}

static List *
get_trigger_cmds(List *trigger_oids)
{
	return get_cmds(trigger_oids, get_trigger_cmd);
}

static List *
get_rule_cmds(List *rule_oids)
{
	return get_cmds(rule_oids, get_rule_cmd);
}

static void
deparse_columns(StringInfo stmt, Relation rel)
{
	int att_idx;
	TupleDesc rel_desc = RelationGetDescr(rel);
	TupleConstr *constraints = rel_desc->constr;

	for (att_idx = 0; att_idx < rel_desc->natts; att_idx++)
	{
		int dim_idx;
		Form_pg_attribute attr = TupleDescAttr(rel_desc, att_idx);

		if (attr->attisdropped)
			continue;

		appendStringInfo(stmt,
						 "\"%s\" %s",
						 NameStr(attr->attname),
						 format_type_with_typemod(attr->atttypid, attr->atttypmod));

		if (attr->attnotnull)
			appendStringInfoString(stmt, " NOT NULL");

		if (OidIsValid(attr->attcollation))
			appendStringInfo(stmt, " COLLATE \"%s\"", get_collation_name(attr->attcollation));

		if (attr->atthasdef)
		{
			int co_idx;

			for (co_idx = 0; co_idx < constraints->num_defval; co_idx++)
			{
				AttrDefault attr_def = constraints->defval[co_idx];

				if (attr->attnum == attr_def.adnum)
				{
					char *attr_default =
						TextDatumGetCString(DirectFunctionCall2(pg_get_expr,
																CStringGetTextDatum(attr_def.adbin),
																ObjectIdGetDatum(rel->rd_id)));

					appendStringInfo(stmt, " DEFAULT %s", attr_default);
					break;
				}
			}
		}

		for (dim_idx = 1; dim_idx < attr->attndims; dim_idx++)
			appendStringInfoString(stmt, "[]");

		if (att_idx != (rel_desc->natts - 1))
			appendStringInfoString(stmt, ", ");
	}
}

typedef struct ConstraintContext
{
	List *constraints;
	List **constraint_indexes;
} ConstraintContext;

static ConstraintProcessStatus
add_constraint(HeapTuple constraint_tuple, void *ctx)
{
	ConstraintContext *cc = ctx;
	Form_pg_constraint constraint = (Form_pg_constraint) GETSTRUCT(constraint_tuple);
	Oid constroid;

	if (OidIsValid(constraint->conindid))
		*cc->constraint_indexes = lappend_oid(*cc->constraint_indexes, constraint->conindid);
#if PG12_GE
	constroid = constraint->oid;
#else
	constroid = HeapTupleGetOid(constraint_tuple);
#endif
	cc->constraints = lappend_oid(cc->constraints, constroid);
	return CONSTR_PROCESSED;
}

static List *
get_constraint_oids(Oid relid, List **constraint_indexes)
{
	ConstraintContext cc = {
		.constraints = NIL,
		.constraint_indexes = constraint_indexes,
	};

	ts_constraint_process(relid, add_constraint, &cc);

	return cc.constraints;
}

static List *
get_index_oids(Relation rel, List *exclude_indexes)
{
	List *indexes = NIL;
	ListCell *cell;

	foreach (cell, RelationGetIndexList(rel))
	{
		Oid indexid = lfirst_oid(cell);

		if (!list_member_oid(exclude_indexes, indexid))
			indexes = lappend_oid(indexes, indexid);
	}
	return indexes;
}

/*
 *  Specifically exclude the hypertable insert blocker from this list.  A table which was recreated
 * with that trigger present would not be able to made into a hypertable.
 */
static List *
get_trigger_oids(Relation rel)
{
	List *triggers = NIL;

	if (rel->trigdesc != NULL)
	{
		int i;

		for (i = 0; i < rel->trigdesc->numtriggers; i++)
		{
			const Trigger trigger = rel->trigdesc->triggers[i];

			if (!trigger.tgisinternal && strcmp(trigger.tgname, INSERT_BLOCKER_NAME) != 0)
				triggers = lappend_oid(triggers, trigger.tgoid);
		}
	}
	return triggers;
}

static List *
get_rule_oids(Relation rel)
{
	List *rules = NIL;

	if (rel->rd_rules != NULL)
	{
		int i;

		for (i = 0; i < rel->rd_rules->numLocks; i++)
		{
			const RewriteRule *rule = rel->rd_rules->rules[i];

			rules = lappend_oid(rules, rule->ruleId);
		}
	}
	return rules;
}

static void
validate_relation(Relation rel)
{
	if (rel->rd_rel->relkind != RELKIND_RELATION)
		ereport(ERROR,
				(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
				 errmsg("given relation is not an ordinary table")));

	if (rel->rd_rel->relpersistence == RELPERSISTENCE_TEMP)
		ereport(ERROR,
				(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
				 errmsg("temporary table is not supported")));

	if (rel->rd_rel->relrowsecurity)
		ereport(ERROR,
				(errcode(ERRCODE_FEATURE_NOT_SUPPORTED), errmsg("row security is not supported")));
}

TableInfo *
deparse_create_table_info(Oid relid)
{
	List *exclude_indexes = NIL;
	TableInfo *table_info = palloc0(sizeof(TableInfo));
	Relation rel = relation_open(relid, AccessShareLock);

	if (rel == NULL)
		ereport(ERROR, (errmsg("relation with id %d not found", relid)));

	validate_relation(rel);

	table_info->relid = relid;
	table_info->constraints = get_constraint_oids(relid, &exclude_indexes);
	table_info->indexes = get_index_oids(rel, exclude_indexes);
	table_info->triggers = get_trigger_oids(rel);
	table_info->rules = get_rule_oids(rel);
	relation_close(rel, AccessShareLock);
	return table_info;
}

TableDef *
deparse_get_tabledef(TableInfo *table_info)
{
	StringInfo create_table = makeStringInfo();
	StringInfo set_schema = makeStringInfo();
	TableDef *table_def = palloc0(sizeof(TableDef));
	Relation rel = relation_open(table_info->relid, AccessShareLock);
	Oid tablespace;

	appendStringInfo(set_schema,
					 "SET SCHEMA '%s';",
					 quote_identifier(get_namespace_name(rel->rd_rel->relnamespace)));
	table_def->schema_cmd = set_schema->data;

	appendStringInfoString(create_table, "CREATE");
	if (rel->rd_rel->relpersistence == RELPERSISTENCE_UNLOGGED)
		appendStringInfoString(create_table, " UNLOGGED");
	appendStringInfoString(create_table, " TABLE");

	appendStringInfo(create_table,
					 " \"%s\".\"%s\" (",
					 get_namespace_name(rel->rd_rel->relnamespace),
					 NameStr(rel->rd_rel->relname));

	deparse_columns(create_table, rel);

	appendStringInfoChar(create_table, ')');

	tablespace = get_rel_tablespace(table_info->relid);
	if (tablespace != InvalidOid)
		appendStringInfo(create_table, " TABLESPACE %s", get_tablespace_name(tablespace));

	appendStringInfoChar(create_table, ';');
	table_def->create_cmd = create_table->data;

	table_def->constraint_cmds = get_constraint_cmds(table_info->constraints);
	table_def->index_cmds = get_index_cmds(table_info->indexes);
	table_def->trigger_cmds = get_trigger_cmds(table_info->triggers);
	table_def->rule_cmds = get_rule_cmds(table_info->rules);

	relation_close(rel, AccessShareLock);
	return table_def;
}

List *
deparse_get_tabledef_commands(Oid relid)
{
	TableInfo *table_info = deparse_create_table_info(relid);
	TableDef *table_def = deparse_get_tabledef(table_info);

	return deparse_get_tabledef_commands_from_tabledef(table_def);
}

List *
deparse_get_tabledef_commands_from_tabledef(TableDef *table_def)
{
	List *cmds = NIL;

	cmds = lappend(cmds, (char *) table_def->schema_cmd);
	cmds = lappend(cmds, (char *) table_def->create_cmd);
	cmds = list_concat(cmds, table_def->constraint_cmds);
	cmds = list_concat(cmds, table_def->index_cmds);
	cmds = list_concat(cmds, table_def->trigger_cmds);
	cmds = list_concat(cmds, table_def->rule_cmds);
	return cmds;
}

const char *
deparse_get_tabledef_commands_concat(Oid relid)
{
	StringInfo tabledef = makeStringInfo();
	ListCell *cell;

	foreach (cell, deparse_get_tabledef_commands(relid))
		appendStringInfoString(tabledef, lfirst(cell));

	return tabledef->data;
}

static const char *
deparse_get_add_dimension_command(Hypertable *ht, Dimension *dimension)
{
	StringInfo dim_cmd = makeStringInfo();

	appendStringInfo(dim_cmd,
					 "SELECT * FROM %s.add_dimension(%s, %s, ",
					 quote_identifier(ts_extension_schema_name()),
					 quote_literal_cstr(
						 quote_qualified_identifier(get_namespace_name(
														get_rel_namespace(ht->main_table_relid)),
													get_rel_name(ht->main_table_relid))),
					 quote_literal_cstr(NameStr(dimension->fd.column_name)));

	if (dimension->type == DIMENSION_TYPE_CLOSED)
		appendStringInfo(dim_cmd,
						 "number_partitions => %d, partitioning_func => %s);",
						 dimension->fd.num_slices,
						 quote_literal_cstr(
							 quote_qualified_identifier(NameStr(
															dimension->fd.partitioning_func_schema),
														NameStr(dimension->fd.partitioning_func))));
	else
		appendStringInfo(dim_cmd, "chunk_time_interval => %ld);", dimension->fd.interval_length);

	return dim_cmd->data;
}

DeparsedHypertableCommands *
deparse_get_distributed_hypertable_create_command(Hypertable *ht)
{
	Hyperspace *space = ht->space;
	Dimension *time_dim = &space->dimensions[0];
	StringInfo hypertable_cmd = makeStringInfo();
	DeparsedHypertableCommands *result = palloc(sizeof(DeparsedHypertableCommands));

	appendStringInfo(hypertable_cmd,
					 "SELECT * FROM %s.create_hypertable(%s",
					 quote_identifier(ts_extension_schema_name()),
					 quote_literal_cstr(
						 quote_qualified_identifier(get_namespace_name(
														get_rel_namespace(ht->main_table_relid)),
													get_rel_name(ht->main_table_relid))));

	appendStringInfo(hypertable_cmd,
					 ", time_column_name => %s",
					 quote_literal_cstr(NameStr(time_dim->fd.column_name)));

	if (time_dim->fd.partitioning_func.data[0] != '\0')
		appendStringInfo(hypertable_cmd,
						 ", time_partitioning_func => %s",
						 quote_literal_cstr(
							 quote_qualified_identifier(NameStr(
															time_dim->fd.partitioning_func_schema),
														NameStr(time_dim->fd.partitioning_func))));

	appendStringInfo(hypertable_cmd,
					 ", associated_schema_name => %s",
					 quote_literal_cstr(NameStr(ht->fd.associated_schema_name)));
	appendStringInfo(hypertable_cmd,
					 ", associated_table_prefix => %s",
					 quote_literal_cstr(NameStr(ht->fd.associated_table_prefix)));

	appendStringInfo(hypertable_cmd, ", chunk_time_interval => %ld", time_dim->fd.interval_length);

	if (OidIsValid(ht->chunk_sizing_func))
	{
		appendStringInfo(hypertable_cmd,
						 ", chunk_sizing_func => %s",
						 quote_literal_cstr(
							 quote_qualified_identifier(NameStr(ht->fd.chunk_sizing_func_schema),
														NameStr(ht->fd.chunk_sizing_func_name))));
		appendStringInfo(hypertable_cmd, ", chunk_target_size => '%ld'", ht->fd.chunk_target_size);
	}

	/*
	 * Backend is assumed to not have any preexisting conflicting table or hypertable.  Any default
	 * indicies will have already been created by the frontend.
	 */
	appendStringInfoString(hypertable_cmd, ", if_not_exists => FALSE");
	appendStringInfoString(hypertable_cmd, ", migrate_data => FALSE");
	appendStringInfoString(hypertable_cmd, ", create_default_indexes => FALSE");
	appendStringInfoString(hypertable_cmd, ", replication_factor => 0");

	appendStringInfoString(hypertable_cmd, ");");

	result->table_create_command = hypertable_cmd->data;
	result->dimension_add_commands = NIL;

	if (space->num_dimensions > 1)
	{
		int i;

		for (i = 1; i < space->num_dimensions; i++)
			result->dimension_add_commands =
				lappend(result->dimension_add_commands,
						(char *) deparse_get_add_dimension_command(ht, &space->dimensions[i]));
	}

	return result;
}
