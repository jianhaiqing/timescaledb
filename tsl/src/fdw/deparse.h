/*
 * This file and its contents are licensed under the Timescale License.
 * Please see the included NOTICE for copyright information and
 * LICENSE-TIMESCALE for a copy of the license.
 */
#ifndef TIMESCALEDB_TSL_FDW_DEPARSE_H
#define TIMESCALEDB_TSL_FDW_DEPARSE_H

#include <postgres.h>

extern void deparseInsertSql(StringInfo buf, RangeTblEntry *rte, Index rtindex, Relation rel,
							 List *targetAttrs, bool doNothing, List *returningList,
							 List **retrieved_attrs);

extern void deparseUpdateSql(StringInfo buf, RangeTblEntry *rte, Index rtindex, Relation rel,
							 List *targetAttrs, List *returningList, List **retrieved_attrs);

extern void deparseDeleteSql(StringInfo buf, RangeTblEntry *rte, Index rtindex, Relation rel,
							 List *returningList, List **retrieved_attrs);

extern bool is_foreign_expr(PlannerInfo *root, RelOptInfo *baserel, Expr *expr);

extern void classifyConditions(PlannerInfo *root, RelOptInfo *baserel, List *input_conds,
							   List **remote_conds, List **local_conds);

extern List *build_tlist_to_deparse(RelOptInfo *foreignrel);

extern void deparseSelectStmtForRel(StringInfo buf, PlannerInfo *root, RelOptInfo *rel, List *tlist,
									List *remote_conds, List *pathkeys, bool is_subquery,
									List **retrieved_attrs, List **params_list);

extern const char *get_jointype_name(JoinType jointype);
extern void deparseStringLiteral(StringInfo buf, const char *val);
extern void deparseAnalyzeSizeSql(StringInfo buf, Relation rel);
extern void deparseAnalyzeSql(StringInfo buf, Relation rel, List **retrieved_attrs);

#endif /* TIMESCALEDB_TSL_FDW_DEPARSE_H */
