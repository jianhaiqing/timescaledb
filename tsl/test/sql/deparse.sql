-- This file and its contents are licensed under the Timescale License.
-- Please see the included NOTICE for copyright information and
-- LICENSE-TIMESCALE for a copy of the license.

-- We compare information(\d+) about manually created tables with the ones that were recreated using deparse_table command.
-- There should be no diff.

\set ECHO errors
\ir include/deparse_func.sql

SELECT format('%s/results/deparse_create.out', :'TEST_OUTPUT_DIR') AS "CREATE_OUT",
       format('%s/results/deparse_recreate.out', :'TEST_OUTPUT_DIR') AS "RECREATE_OUT"
\gset
SELECT format('\! diff %s %s', :'CREATE_OUT', :'RECREATE_OUT') as "DIFF_CMD"
\gset

\ir include/deparse_create.sql

\o :CREATE_OUT
\d+ "public".*

\ir include/deparse_recreate.sql
\o :RECREATE_OUT
\d+ "public".*

\o

:DIFF_CMD

SELECT 'TABLE DEPARSE TEST DONE';

\set ECHO all
-- test drop_chunks function deparsing
SELECT * FROM tsl_test_deparse_drop_chunks('2019-01-01'::timestamptz, 'test_table', 'test_schema', cascade => false, verbose => true);
SELECT * FROM tsl_test_deparse_drop_chunks(interval '1 day', table_name => 'weird nAme\\#^.', cascade_to_materializations => true, cascade => true);
SELECT * FROM tsl_test_deparse_drop_chunks(newer_than => 12345);
SELECT * FROM tsl_test_deparse_drop_chunks(older_than => interval '2 years', newer_than => '2015-01-01'::timestamp);
