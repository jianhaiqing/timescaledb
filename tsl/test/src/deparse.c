/*
 * This file and its contents are licensed under the Timescale License.
 * Please see the included NOTICE for copyright information and
 * LICENSE-TIMESCALE for a copy of the license.
 */
#include <postgres.h>
#include <utils/builtins.h>
#include "export.h"
#include "deparse.h"

TS_FUNCTION_INFO_V1(tsl_test_get_tabledef);
TS_FUNCTION_INFO_V1(tsl_test_deparse_drop_chunks);

Datum
tsl_test_get_tabledef(PG_FUNCTION_ARGS)
{
	Oid relid = PG_GETARG_OID(0);
	const char *cmd = deparse_get_tabledef_commands_concat(relid);
	PG_RETURN_TEXT_P(cstring_to_text(cmd));
}

Datum
tsl_test_deparse_drop_chunks(PG_FUNCTION_ARGS)
{
	Name table_name = PG_ARGISNULL(1) ? NULL : PG_GETARG_NAME(1);
	Name schema_name = PG_ARGISNULL(2) ? NULL : PG_GETARG_NAME(2);
	Datum older_than_datum = PG_GETARG_DATUM(0);
	Datum newer_than_datum = PG_GETARG_DATUM(4);
	Oid older_than_type = PG_ARGISNULL(0) ? InvalidOid : get_fn_expr_argtype(fcinfo->flinfo, 0);
	Oid newer_than_type = PG_ARGISNULL(4) ? InvalidOid : get_fn_expr_argtype(fcinfo->flinfo, 4);
	bool cascade = PG_GETARG_BOOL(3);
	bool verbose = PG_ARGISNULL(5) ? false : PG_GETARG_BOOL(5);
	bool cascades_to_materializations = PG_ARGISNULL(6) ? false : PG_GETARG_BOOL(6);
	const char *sql_cmd;

	sql_cmd = deparse_drop_chunks_func(table_name,
									   schema_name,
									   older_than_datum,
									   newer_than_datum,
									   older_than_type,
									   newer_than_type,
									   cascade,
									   cascades_to_materializations,
									   verbose);

	PG_RETURN_TEXT_P(cstring_to_text(sql_cmd));
}
