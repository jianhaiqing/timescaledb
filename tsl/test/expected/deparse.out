-- This file and its contents are licensed under the Timescale License.
-- Please see the included NOTICE for copyright information and
-- LICENSE-TIMESCALE for a copy of the license.
-- We compare information(\d+) about manually created tables with the ones that were recreated using deparse_table command.
-- There should be no diff.
\set ECHO errors
        ?column?         
-------------------------
 TABLE DEPARSE TEST DONE
(1 row)

-- test drop_chunks function deparsing
SELECT * FROM tsl_test_deparse_drop_chunks('2019-01-01'::timestamptz, 'test_table', 'test_schema', cascade => false, verbose => true);
                                                                                                               tsl_test_deparse_drop_chunks                                                                                                                
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 SELECT * FROM public.drop_chunks(older_than => 'Tue Jan 01 00:00:00 2019 PST'::timestamptz, newer_than => NULL, table_name => 'test_table', schema_name => 'test_schema', cascade => 'false', cascade_to_materializations => 'false', verbose => 'true');
(1 row)

SELECT * FROM tsl_test_deparse_drop_chunks(interval '1 day', table_name => 'weird nAme\\#^.', cascade_to_materializations => true, cascade => true);
                                                                                                  tsl_test_deparse_drop_chunks                                                                                                   
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 SELECT * FROM public.drop_chunks(older_than => '@ 1 day'::interval, newer_than => NULL, table_name => E'weird nAme\\\\#^.', schema_name => NULL, cascade => 'true', cascade_to_materializations => 'true', verbose => 'false');
(1 row)

SELECT * FROM tsl_test_deparse_drop_chunks(newer_than => 12345);
                                                                                        tsl_test_deparse_drop_chunks                                                                                         
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 SELECT * FROM public.drop_chunks(older_than => NULL, newer_than => '12345'::int4, table_name => NULL, schema_name => NULL, cascade => 'false', cascade_to_materializations => 'false', verbose => 'false');
(1 row)

SELECT * FROM tsl_test_deparse_drop_chunks(older_than => interval '2 years', newer_than => '2015-01-01'::timestamp);
                                                                                                             tsl_test_deparse_drop_chunks                                                                                                             
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 SELECT * FROM public.drop_chunks(older_than => '@ 2 years'::interval, newer_than => 'Thu Jan 01 00:00:00 2015'::timestamp, table_name => NULL, schema_name => NULL, cascade => 'false', cascade_to_materializations => 'false', verbose => 'false');
(1 row)

